walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,170 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Meeting cannot last longer than 24 hours
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) return Arrays.asList();
+
+    Collection<TimeRange> availableTimesWithOptionalAttendees = queryHelper(events, request, true);
+
+    // If there are no available times with all (mandatory and optional attendees) then only consider mandatory 
+    if(availableTimesWithOptionalAttendees.size() == 0 && request.getAttendees().size() > 0) {
+      return queryHelper(events, request, false);
+    }
+
+    return availableTimesWithOptionalAttendees;
+  }
+
+  /**
+   * Helper method for the query method which finds the available times to meet given a list of events and a MeetingRequest,
+   * depending on whether optional attendees are considered or not.
+   * @param events the list of events to take into consideration when finding available times
+   * @param request the meeting request with the meeting attendees and duration
+   * @param considerOptional whether optional attendees are accounted for or not
+   * @return available times for the group to meet
+   *
+   */
+  private Collection<TimeRange> queryHelper(Collection<Event> events, MeetingRequest request, boolean considerOptional) {
+    Collection<String> attendees = new HashSet<String>(request.getAttendees());
+    if(considerOptional) {
+      attendees.addAll(request.getOptionalAttendees());
+    }
+    
+    // Needs to have at least 1 attendee at a meeting
+    if(attendees.size() == 0) return Arrays.asList(TimeRange.WHOLE_DAY);
+
+    // For a time to be unavailable, there must be someone at the event who is in the meeting attendees list
+    List<TimeRange> unavailableTimes = new ArrayList(convertAttendeesAndEventsToUnavailableTimes(events, attendees));
+
+    // If there are no unavailable times, people can meet any time!
+    if(unavailableTimes.size() == 0) return Arrays.asList(TimeRange.WHOLE_DAY);
+
+    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+
+    Collection<TimeRange> unavailableNoOverlaps = mergeTimeRanges(unavailableTimes);
+
+    Collection<TimeRange> availableTimes = findAvailableTimesFromUnavailableTimes(unavailableNoOverlaps);
+
+    Collection<TimeRange> finalAvailableTimes = removeAllTimeRangesLessThanGivenDuration(availableTimes, request.getDuration());
+
+    return finalAvailableTimes;
+  }
+
+  /**
+   * Removes items from a Collection if they are shorter than the specified duration.
+   * @param times a collection of TimeRanges
+   * @param duration the threshold length of the meeting 
+   * @return a new collection with a list of sufficient times at least as long as the given duration
+   */
+  private static Collection<TimeRange> removeAllTimeRangesLessThanGivenDuration(Collection<TimeRange> times, long duration) {
+    Collection<TimeRange> sufficientTimes = new ArrayList<>(); 
```

Comment:
I'm not sure exactly which version of Java you're using for the walkthroughs, but have you used Java Collections and Streams before (they're available in Java 8 and above)? I think this method and a couple below could be good opportunities to try them out (here's a sample tutorial with a few examples: http://tutorials.jenkov.com/java-collections/streams.html).

Reply:
Didn't know about Streams! Thank you!